include/pushmi/traits.h

```diff
@@ -65,6 +65,44 @@ using void_t = void;
 template <class T>
 using remove_cvref_t = std::remove_cv_t<std::remove_reference_t<T>>;
 
+//
+// courtesy of Eric Niebler
+// check_functor does not work for types marked final.
+//
+
+struct check_functor_base {
+    void operator()() {}
+};
+
+template <class T>
+struct check_functor : check_functor_base, T {};
+
+PUSHMI_CONCEPT_DEF(
+  template (class T)
+  (concept NotFunctor) (T),
+    requires(T) (
+      &check_functor<T>::operator()
+    )
+);
+
+PUSHMI_CONCEPT_DEF(
+  template (class T)
+  (concept Functor) (T),
+    not NotFunctor<T>
+);
+
+PUSHMI_CONCEPT_DEF(
+  template (class T)
+  (concept Function) (T),
+    std::is_function<T>::value
+);
+
+PUSHMI_CONCEPT_DEF(
+  template (class T)
+  (concept Callable) (T),
+    Functor<T> || Function<T>
+);
+
```

Comment:
not Receiver is the alternative to Callable.. when the negative test was not working, I went with a positive test..

Reply:
> > The Callable concept has no syntax and no semantics
> 
> I would like to hear more. IMO Callable is constraining on both syntax and semantics. Callable requires that a value of type T supports the C++ function call syntax (without constraining the arity, the parameter types or the result type). Callable requires that a value of type T supports the C++ function call semantic for some, as yet unspecified, set of parameters.

When evaluating whether an imagined concept actually has meaningful syntax and semantics, consider what a hypothetical algorithm can _do_ with a type constrained with that concept.

In other words, what can following function `foo` do with the object `o`?

```c++
template <Callable T>
void foo( T o )
{
  // what can foo do with o?
}
```

Given the definition of `Callable`, the _only_ thing it can do is to forward `o` to one of two functions: one which is constrained with `Function` and the other which is constrained with `Functor`. The function constrained with `Functor` can then do _nothing_, absolutely nothing, with the object `o`. That is not a real concept.

If you are only using the concept machinery as a metaprogramming aid as an implementation detail, then it should be hidden in a detail namespace. But it's a pretty lousy implementation because it is easily broken (e.g., by a function object that is marked `final`).

Reply:
*just found this - never hit submit*

If I change the names in the code, and add an implementation, then *Callable* will become meaningful.

```cpp
template <Callable T, Regular P0>
auto bind_first( T o, P0 v )
{
  // Q: what can bind_first do with o?

  // A: bind_first can return a new Callable that stores the Callable o and the parameter v 
  return [o, v](auto... pn) /*require Invocable<o, P0, decltype(pn)...>*/ { return o(v, pn...); };
}
```

It is important for bind_first to know that `o` is *Callable* so that it knows that it is meaningful to return a *Callable*.

*Functor* requiring final is nasty, but the workaround is using an overload wrapper like `on_value()`, this does not add significant overhead for users..

Reply:
>   // Q: what can bind_first do with o?
>
>  // A: bind_first can return a new Callable that stores the Callable o and the parameter v 

Ironically it can't because Callable doesn't require CopyConstructible. That aside, `bind_first` is overconstrained because (a) it only ever uses the copyability of `T`, and (b) because it rejects (for no reason that benefits the user) function object types that are marked `final`. It also rejects class types that are callable because they have an implicit conversion to a function pointer. (This is seriously a thing.)

You haven't convinced me that this is a real concept. I suspect you can't.